code/modules/tgui_input/say_modal/tgui_say_modal.dm

```diff
@@ -0,0 +1,135 @@
+/** Assigned say modal of the client */
+/client/var/datum/tgui_say/tgui_say
+
+/**
+ * Creates a JSON encoded message to open TGUI say modals properly.
+ *
+ * Arguments:
+ * channel - The channel to open the modal in.
+ * Returns:
+ * string - A JSON encoded message to open the modal.
+ */
+/client/proc/tgui_say_create_open_command(channel)
+	var/message = TGUI_CREATE_MESSAGE("open", list(
+		channel = channel,
+	))
+	return "\".output tgui_say.browser:update [message]\""
+
+/**
+ * The tgui say modal. This initializes an input window which hides until
+ * the user presses one of the speech hotkeys. Once something is entered, it will
+ * delegate the speech to the proper channel.
+ */
+/datum/tgui_say
+	/// The user who opened the window
+	var/client/client
+	/// Injury phrases to blurt out
+	var/static/list/hurt_phrases = list("AGH!", "OUF!", "OOF!", "AUGH!", "OW!", "URGH!", "ACK!")
+	/// The modal window
+	var/datum/tgui_window/window
+	/// Boolean for whether the tgui_say was opened by the user.
+	var/window_open
+
+/** Creates the new input window to exist in the background. */
+/datum/tgui_say/New(client/client, id)
+	src.client = client
+	window = new(client, id)
+	winset(client, "tgui_say", "size=1,1;is-visible=0;")
+	window.subscribe(src, PROC_REF(on_message))
+	window.is_browser = TRUE
+
+/**
+ * After a brief period, injects the scripts into
+ * the window to listen for open commands.
+ */
+/datum/tgui_say/proc/initialize()
+	set waitfor = FALSE
+	// Sleep to defer initialization to after client constructor
+	sleep(3 SECONDS)
+	window.initialize(
+			strict_mode = TRUE,
+			fancy = TRUE,
+			inline_css = file("tgui/public/tgui-say.bundle.css"),
+			inline_js = file("tgui/public/tgui-say.bundle.js"),
+	);
+
+/**
+ * Ensures nothing funny is going on window load.
+ * Minimizes the window, sets max length, closes all
+ * typing and thinking indicators. This is triggered
+ * as soon as the window sends the "ready" message.
+ */
+/datum/tgui_say/proc/load()
+	window_open = FALSE
+	winset(client, "tgui_say", "pos=848,500;size=231,30;is-visible=0;")
+	window.send_message("props", list(
+		lightMode = (client.prefs.toggles2 & PREFTOGGLE_2_ENABLE_TGUI_SAY_LIGHT_MODE),
+		maxLength = MAX_MESSAGE_LEN,
+	))
```

Comment:
I think these are keys in an associative list, right? If so, we should clarify that
```suggestion
	window.send_message("props", list(
		"lightMode" = (client.prefs.toggles2 & PREFTOGGLE_2_ENABLE_TGUI_SAY_LIGHT_MODE),
		"maxLength" = MAX_MESSAGE_LEN,
	))
```